#8086 Layout changes and other improvements

alexander-fedorenko opened the pull request:
## Description
Done
- Annotations moved to the left sidebar & TOC;
- Print plugin updated to support optional print of additional layers;
- Sidebar menu plugin.
- Updating another plugins for new SidebarMenu container.
- Finalize update of sidebar menu members.
- Update layout properly when any of the right panels is open.
- Handle the case when sidebar menu has too many elements to fit at its height
- Reduce right panels sizes.
- Updated search bar behavior.
- Tests coverage.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

<!-- add here the ReadTheDocs link (if needed) -->

## Issue

**What is the current behavior?**
https://github.com/geosolutions-it/MapStore2/issues/8086
[#150](https://github.com/georchestra/mapstore2-cadastrapp/issues/150)

**What is the new behavior?**
As per description

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [x] Yes, and I documented them in migration notes
 - [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications -->

## Other useful information

offtherailz reviewed (changes requested):
It is cool. Here I report the things I think should be changed or fixed, please @alexander-fedorenko feel free to tell if I do a mistake or if you have any other comments.

- [x] I see (except for annotation, that is similar anyway) that `DockConteiner --> ContainerDimensions --> Dockpanel` is often used, managing size in a similar way too. It  make sense to create a reusable component for this, a sort of SidePanel, documenting the properties accordingly? It looks to me a sort of side panel that adapt it's width. What do you think?
- [x] widgets are not visible anymore in this map. http://localhost:8081/#/viewer/openlayers/10358
- [x] About migration and configuration of context, we may need to think a little how to do it. Maybe a possibility is to make the burger menu visible in context editor, and make it not a dependency of any plugin. So we can remove it for old contexts and re-save, using the side-bar instead (It is just the first idea I have). 

**I suggest to test this approach for this point**
- Make the BurgerMenu plugin visible in context editor
- Make the plugins dependent only by Sidebar, removing every dependency from BurgerMenu
- Make BurgerMenu to have higher priority then the sidebar
- SideBar will remain hidden

This should allow the user to manually edit the contexts by removing manually the BurgerMenu. If instead the BurgerMenu is Present in the list, it will be used instead of the sidebar. 
I think this BurgerMenu priority works if the sidebar, when empty, is not rendered. 
This solution should be fast to be implemented (only configuration in `pluginsConfig.json` and priority changes, plus maybe the empty - hide task for sidebar, if needed), So I suggest to implement and so we can test it. 


- [x] We should remove annotations tool from sidebar, we already have it in the TOC. At least set it as an alternative choise (less priority).
- [x] We need to add migration guidelines for:
    - Replacing the new plugin (and eventually update cotnexts)
    - replace plugin to localConfig 

- [x] Remove the burger menu from the viewer at all and avoid disablePluginIf and monitored state if possible. 

@allyoucanmap can you please check the part of the code dedicated to map layout, that you know better than me (and that may be more delicate for the projects you follow).

allyoucanmap reviewed (changes requested):
@tdipisa @offtherailz @alexander-fedorenko

Below some general thoughts on the UI that could be worked within this task or on separated issue.

The sidebar is working as expected but there are some parts that did not seems balanced with the whole layout:

## Current layout result

map viewer

![image](https://user-images.githubusercontent.com/19175505/165816225-6f54df43-88f4-4781-b158-388667b47ac9.png)

dashboard viewer

![image](https://user-images.githubusercontent.com/19175505/165816435-c5dce064-1cc5-408a-b66e-0c73e96b535a.png)

geostory viewer

![image](https://user-images.githubusercontent.com/19175505/165816521-623410df-5797-45f9-aa13-900201564fbe.png)

Feedback

- the buttons of the sidebar introduce a new size they are not the default size neither the medium one but they introduce a new 40x40px size. Maybe this is the time to reduce the size of the default square buttons.
- there isn't a line of separation between the side bar and the background. Maybe we need a border or a soft shadow
- some icons are duplicated and the will need to be changed (eg: save and save as)
- floating search, map toolbar, toc button and background selector do not have a consistent padding from the view limits
-  geostory and dashboard will still use the background selector so they will have an inconsistent UX. Should we start to think to have a vertical bar there?
- bug on the dahboard/geostory css of omnibar found by @alexander-fedorenko 

## Test

In the below screenshot I tried to use same padding on floating elements on map, add a border/shadow to side panel and align the button sizes with the one of the sidebar with this variables:

Note: probably it's not enough to change this two variables anyway because some other components style could need some additional fixes

```less
@icon-size: 24px;
@square-btn-size: 40px;
```

Probably also the search bar should be reduced in height to match the button size at least in the default view.


- side bar with left border

![image](https://user-images.githubusercontent.com/19175505/165817559-d46250b8-865c-4fbe-917d-ee9224810ac8.png)

- side bar with shadow-soft class

![image](https://user-images.githubusercontent.com/19175505/165817396-aafa603b-e44c-4705-81b5-52e6892e2136.png)

offtherailz commented:
@alexander-fedorenko 
About the 1st point of my review, take in condieration also what @allyoucanmap suggested for the resize-monitor.

I updated the **3rd point** of my review with things disucssed with @allyoucanmap and @tdipisa. Please try to implement the suggested solution so we can test if it is easy and manageble as supposed.

tdipisa commented on web/client/plugins/Print.jsx:
```diff
@@ -608,18 +612,19 @@ export default {
                     error,
                     map,
-                    layers: layers.filter(l => !l.loadingError),
+                    layers: [...layers.filter(l => !l.loadingError), ...(printSpec?.additionalLayers ? additionalLayers.map(l => l.options).filter(l => !l.loadingError) : [])],
```
@alexander-fedorenko please see last point of my comment below for this.

alexander-fedorenko commented on web/client/epics/maplayout.js:
```diff
@@ -124,13 +139,20 @@ export const updateMapLayoutEpic = (action$, store) =>
+                } else {
+                    boundingMapRect[key] = (parseFloat(boundingMapRect[key]) + parseFloat(boundingSidebarRect[key] ?? 0)) + '%';
+                }
+            });
```
That's the workaround to recalculate boundingMapRect according to the values provided via boundingSidebarRect. Unfortunately, only part of the properties is available as numeric values, while others are transformed into percentage amount right away. Else statement parses numbers out of percentage amount, sum them and transforms final value back to percentage.

alexander-fedorenko commented:
Addressed items:
- Not visible widgets on the map http://localhost:8081/#/viewer/openlayers/10358  // can't reproduce, all works fine, maybe some glitch on remote env?
- Contexts migration: 

I've applied proposed approach. Though, it was not straight-forward to hide SidebarMenu. We have two items rendered either in Omnibar or in SidebarMenu. Therefore these items (Home, Login) needs to be rendered in sidebar only in case if BurgerMenu is missing on the page and in Omnibar otherwise. I've applied workaround to hide these items and make Sidebar menu track its own state and hide/show when there are no items to render / at least one item to render.

- Removed annotations from Sidebar.
- BurgerMenu is removed from viewer by removal of entry from desktop -> "BurgerMenu" in localConfig.json. It will remain active on Geostory, Dashboard pages.
- Monitored state for Sidebar is removed;
- Buttons size is adjusted. I checked briefly and corrected styles where it was looking odd, but I might miss something.
- Applied shadow to the sidebar.
- Updated icon for Save tool.
- Paddings of searchbar, map toolbar and background selector should be 5px width now.
- Reverted Omnibar CSS changes, so navbar should be ok on all pages.
- Option to print overlay layers is deactivated by default and renamed. Can be activated by passing following configuration in localConfig:
```
{
  "name": "Print",
  "cfg": {
    "overlayLayersOptions": {
      "enabled": false
    }
  }
}
```

Not addressed:
- New abstract component to replace DockContainer --> ContainerDimensions --> Dockpanel sequence. 
- react-container-dimensions is in use for a few other plugins/components: Floating legend, Geostory / Cascade, Geostory / Sections / Title, Banner, Filters Editor, Adaptive grid etc.  Probably it's good to be replaced, but not in a scope of this PR.
- Migration guidelines;
- Adding annotations to the TOC using injection from another plugin.

allyoucanmap commented:
> TODO: let's use the new size for all buttons (also the size of medium buttons can be changed). @allyoucanmap will explain better this.

@alexander-fedorenko about the medium buttons. In some project  I used to apply this variables also for medium square buttons (eg. TOC toolbar buttons):

```
@icon-size: 24px;
@square-btn-size: 40px;

@icon-size-md: 15px;
@square-btn-medium-size: 30px;
```

alexander-fedorenko commented:
@offtherailz @allyoucanmap Could you please check it, considering that remaining changes won't change anything in UI/UX. Last three items will be available for review tomorrow morning.

alexander-fedorenko commented on web/client/components/TOC/Toolbar.jsx:
```diff
@@ -326,6 +327,26 @@ class Toolbar extends React.Component {
+                            <Glyphicon glyph="pencil" />
+                        </Button>
+                    </OverlayTrigger>
+                    : null}
```
Thanks, implemented this approach

allyoucanmap reviewed (commented):
The layout in general looks good, these are some things probably related to the button size change

- global loader spinner is not centered. I think there is some padding forced somewhere
- the widgets are overlapping the map toolbar

![layout-imp](https://user-images.githubusercontent.com/19175505/166470591-e7f4eec3-8d57-460d-8bc7-122bea01cc8e.png)

- there is a gap between header and content in the dashboard and geostory pages probably due to the padding of the logo

![layout-dahboard-gap](https://user-images.githubusercontent.com/19175505/166470607-872afb8d-85de-4850-a4ec-82d1a0c538fb.png)

![layout-geostory-gap](https://user-images.githubusercontent.com/19175505/166470667-34334502-3926-4eee-b108-3335aa80b251.png)

alexander-fedorenko commented:
Hi @allyoucanmap, these problems are resolved.
With the recent commits:
- Updated widgets plugin, switched from hard-coded offset to tracking of the availability of sidebar like other tools. Same for widgets toolbar at the bottom.
- Widgets were not hidden properly when any of the right panels is open. It should be gone as I've added a flags of open left/right panels into map layout and updated selector that checks if panels are open. This removes need to make a hard-coded comparison in selector and makes it use state. Minus 1 to-do item in code.
- Corrected spinner CSS.
- Logo style on dashboard/geostory are corrected.

offtherailz reviewed (changes requested):
Hi.
I tested the changes and the things with burger menu works as expected. Very good.

 I did some tests more and I noticed these things:

- [ ] The export, save, save as and delete map tools doesn't work. 

https://user-images.githubusercontent.com/1279510/166660538-92acc69a-03bb-455b-9755-4c21d1c725c1.mp4

- [ ] Map template plugin has a small problem in rendering maps (here a context to test https://dev-mapstore.geosolutionsgroup.com/mapstore/#/context/test-nali-2-templates) 
![image](https://user-images.githubusercontent.com/1279510/166661677-12d1dc46-0679-4b03-b25f-6553dbdef62f.png)  Also if I use the toolbar, same problem 
![image](https://user-images.githubusercontent.com/1279510/166661969-0909b1f8-cd42-498b-aaf8-a4dcf287aafb.png)
- [ ] Street view plugin is not visible in the toobar

About the code and doc: 

- [ ] Please document more `ResponsivePanel` (and the internal `DockContainer`). They are used widely and we need to have a clear definition of what  they do and their props. From my understanding it defines a general container structure for a docked on a side that automatically fit the page (or its container) when the screen size is smaller. Correct? @allyoucanmap now we can, maybe in a second time, use the contaner-monitor to replace container dimension in a dedicated issue. 
- [ ] Add to the migration guidelines also the changes to apply to the `pluginsConfig.json`

About widgets:
- [x] Widgets seems to be ok. Maybe you are right, I did some mistake.

allyoucanmap commented:
> @allyoucanmap now we can, maybe in a second time, use the contaner-monitor to replace container dimension in a dedicated issue.

@alexander-fedorenko @offtherailz yes, we can do that another moment

offtherailz commented on docs/developer-guide/mapstore-migration-guide.md:
````diff
@@ -107,6 +107,45 @@ Following actions need to be applied to make a switch:
 
 #### Updating contexts to use Sidebar Menu
+Contents of your `pluginsConfig.json` need to be reviewed to allow usage of new "SidebarMenu" in contexts.
+- Find "BurgerMenu" plugin confuguration in `pluginsConfig.json` and remove `"hidden": true` line from it:
+```json
+    {
+  "name": "BurgerMenu",
+  "glyph": "menu-hamburger",
+  "title": "plugins.BurgerMenu.title",
+  "description": "plugins.BurgerMenu.description",
+  "dependencies": [
+    "OmniBar"
+  ]
+}
+```
+- Add "SidebarMenu" entry to the "plugins" array:
+```json
+{
+    "plugins": [
+        ...
+        {
+            "name": "SidebarMenu",
+            "hidden": true
+        }
+        ...
+    ]
+}
+```
+- Go through all plugins definitions and replace "BurgerMenu" dependency with "SidebarMenu", e.g.:
+```json
+    {
+      "name": "MapExport",
+      "glyph": "download",
+      "title": "plugins.MapExport.title",
+      "description": "plugins.MapExport.description",
+      "dependencies": [
+        "SidebarMenu"
+      ]
+    }
+```
````
There are some error in rendring of documentation
(see here:  https://mapstore--8085.org.readthedocs.build/en/8085/developer-guide/mapstore-migration-guide/#replacing-burgermenu-with-sidebarmenu  )

Probably there is no space before the list element ( ` - Find Burger menu ....`)

````suggestion

Contents of your `pluginsConfig.json` need to be reviewed to allow usage of new "SidebarMenu" in contexts.

- Find "BurgerMenu" plugin confuguration in `pluginsConfig.json` and remove `"hidden": true` line from it:

```json
    {
  "name": "BurgerMenu",
  "glyph": "menu-hamburger",
  "title": "plugins.BurgerMenu.title",
  "description": "plugins.BurgerMenu.description",
  "dependencies": [
    "OmniBar"
  ]
}
```

- Add `"SidebarMenu"` entry to the "plugins" array:

```json
{
    "plugins": [
        ...
        {
            "name": "SidebarMenu",
            "hidden": true
        }
        ...
    ]
}
```

- Go through all plugins definitions and replace `"BurgerMenu"` dependency with `"SidebarMenu"`, e.g.:

```json
    {
      "name": "MapExport",
      "glyph": "download",
      "title": "plugins.MapExport.title",
      "description": "plugins.MapExport.description",
      "dependencies": [
        "SidebarMenu"
      ]
    }
```

- Also the `StreetView` plugin needs to depend from `SidebarMenu`. 

```json
{
      "name": "StreetView",
      "glyph": "road",
      "title": "plugins.StreetView.title",
      "description": "plugins.StreetView.description",
      "dependencies": ["Toolbar"]
      "dependencies": [
        "SidebarMenu"
      ]
}
```

````

alexander-fedorenko replied:
Thanks! I've applied your changes with a fix to line 208 above

offtherailz commented on web/client/configs/pluginsConfig.json:
```diff
@@ -569,27 +590,40 @@
       "description": "plugins.StreetView.description",
-      "dependencies": ["Toolbar"]
+      "dependencies": [
+        "Toolbar"
```
```suggestion
        "SidebarMenu"
```

offtherailz reviewed (changes requested):
I sent a couple of fixes for documentation and `pluginsConfig.json`
- One for the pluginsConfig that used the toolbar dependency, but in fact it doesn't use it. So I changed it to `SidebarMenu` (otherwise a context with only streetview will not allow to open street view)
- Another for formatting the doc correctly and document the change above

alexander-fedorenko commented:
Tested changes with geOrchestra, went through all the points mentioned in PR reviews, all looks good. The only thing I was not able to test is context creation workflow (unfortunately I have no access, header/login tool is missing).

tdipisa commented:
@ElenaGallo please extensively test this in DEV as soon as deployed.

ElenaGallo commented:
Hi @alexander-fedorenko testing I found these problems:

1_ The catalog, export and import buttons are no longer present in the configuration map viewer in the context wizard

https://user-images.githubusercontent.com/56537133/166938843-f7000c45-b812-4264-b649-4aa34dc9a5c7.mov

2_ The Burger Menu is still present in the Inline geostory map configuration option and the catalog is not displayed correctly

https://user-images.githubusercontent.com/56537133/166938849-6e5d613e-9103-4ec1-9a0b-1547a6541427.mov

3_ The 3D navigator is covered by the home sidebar
![3](https://user-images.githubusercontent.com/56537133/166938870-1c4402cf-eef7-468e-b031-704bbb32b9a6.png)

tdipisa commented:
@ElenaGallo 

> 2_ The Burger Menu is still present in the Inline geostory map configuration option and the catalog is not displayed correctly

it is not expected to have the sidebar in the geostore advanced editor

ElenaGallo commented:
@tdipisa ok, but i think with the panel size change, the catalog no longer displays correctly (the catalog panel cannot be closed). See the gif:

https://user-images.githubusercontent.com/56537133/167098099-e7ee227f-25d9-440e-8c65-eb3056d4077b.mov

tdipisa commented:
@ElenaGallo ok for this:
>ok, but i think with the panel size change, the catalog no longer displays correctly (the catalog panel cannot be closed).

ElenaGallo commented:
@tdipisa ok, but the catalog panel is not displayed correctly (cannot be closed). See the gif:

https://user-images.githubusercontent.com/56537133/167098545-8bb56324-eb64-4470-ab12-68880cd54ac6.mov
